packages/common-utils/src/renderChartConfig.ts

```diff
@@ -840,6 +840,129 @@ function renderFill(
+  };
+
+  if (aggFn !== 'quantile' || level == null) {
+    throw new Error('quantile must be specified for histogram metrics');
+  }
+
+  // Render the where clause to limit data selection on the source CTE but also search forward/back one
+  // bucket window to ensure that there is enough data to compute a reasonable value on the ends of the
+  // series.
+  const cteChartConfig = {
+    ...chartConfig,
+    from: {
+      ...from,
+      tableName: metricTables[MetricsDataType.Histogram],
+    },
+    filters: [
+      ...(filters ?? []),
+      {
+        type: 'sql',
+        condition: `MetricName = '${metricName}'`,
+      },
+    ],
+    includedDataInterval:
+      chartConfig.granularity === 'auto' && Array.isArray(chartConfig.dateRange)
+        ? convertDateRangeToGranularityString(chartConfig.dateRange, 60)
+        : chartConfig.granularity,
+  } as ChartConfigWithOptDateRangeEx;
+
+  const timeBucketSelect = isUsingGranularity(cteChartConfig)
+    ? timeBucketExpr({
+        interval: cteChartConfig.granularity,
+        timestampValueExpression: cteChartConfig.timestampValueExpression,
+        dateRange: cteChartConfig.dateRange,
+      })
+    : chSql``;
+  const where = await renderWhere(cteChartConfig, metadata);
+  const groupBy = (await renderGroupBy(cteChartConfig, metadata)) || chSql``;
+  const limit = renderLimit(chartConfig);
+
+  return {
+    connection: restChartConfig.connection,
+    with: [
+      {
+        name: 'Points',
+        sql: chSql`
+          SELECT
+            ${timeBucketSelect.sql ? chSql`${timeBucketSelect},` : ''}
+            last_value(HistogramTable.Count) AS Count,
+            last_value(HistogramTable.BucketCounts) AS BucketCounts,
+            last_value(HistogramTable.ExplicitBounds) AS ExplicitBounds,
+            last_value(HistogramTable.Sum) AS Sum,
+            last_value(HistogramTable.AggregationTemporality) AS AggregationTemporality,
+            cityHash64(mapConcat(ScopeAttributes, ResourceAttributes, Attributes)) AS AttributesHash,
+            toUInt64( ${{ Float64: level }} * Count) AS Rank,
+            arrayCumSum(BucketCounts) as CumCounts,
+            arrayFirstIndex(x -> if(x > Rank, 1, 0), CumCounts) AS BucketLowIdx,
+            CASE
+              WHEN Count = 0 THEN 0
+              WHEN Count = 1 THEN Sum
+              WHEN BucketLowIdx = 1 THEN ExplicitBounds[BucketLowIdx]
+              WHEN BucketLowIdx = length(BucketCounts) THEN ExplicitBounds[length(ExplicitBounds)]
+              ELSE
+                ExplicitBounds[BucketLowIdx - 1] +
+                  ((Rank - ExplicitBounds[BucketLowIdx - 1]) / (BucketCounts[BucketLowIdx])) * (ExplicitBounds[BucketLowIdx] - ExplicitBounds[BucketLowIdx - 1])
+            END AS Point
+          FROM ${renderFrom({ from: { ...from, tableName: metricTables[MetricsDataType.Histogram] } })} HistogramTable
+          WHERE ${where}
+          GROUP BY ${concatChSql(',', [chSql`AttributesHash`, groupBy])}
```

Comment:
we run into problems here if we try to group by granularity before calculating rate, usually when we start setting granularity to be something like looking back 14d and then we'll do something like 12h granularity, and with rates only having 12h granularity we run into huge errors with counter resets making the metric virtually useless.

In v1 I believe we've had success with doing our rate calculations first and then doing the percentile interpolation, I'm wondering what's the issue we're running into here, is it for some reason the rate calculations on arrays are really memory expensive?